File: StackStackStack/StackStackStack/ImageEncoder.swift
import Foundation
import ImageIO
import UniformTypeIdentifiers
import CoreGraphics

enum ImageEncoderError: Error { case contextFailed, destinationFailed, finalizeFailed }

enum ImageEncoder {
    /// `String`-backed: raw values are persisted (AppSettings + StackRecord.format) — renaming a
    /// case silently breaks stored preferences and library records.
    enum Format: String, Sendable, Equatable {
        case jpeg, heic
        var utType: UTType { self == .jpeg ? .jpeg : .heic }
        /// The file extension library files use for this format.
        var fileExtension: String { self == .jpeg ? "jpg" : "heic" }
    }

    /// EXIF metadata written into the encoded result.  All fields are optional; nil = omit.
    struct ExifMetadata {
        /// ISO speed (maps to kCGImagePropertyExifISOSpeedRatings).
        var iso: Double?
        /// Exposure time in seconds (maps to kCGImagePropertyExifExposureTime).
        var shutterSeconds: Double?
        /// Capture timestamp (maps to kCGImagePropertyExifDateTimeOriginal).
        var capturedAt: Date?

        init(iso: Double? = nil, shutterSeconds: Double? = nil, capturedAt: Date? = nil) {
            self.iso = iso
            self.shutterSeconds = shutterSeconds
            self.capturedAt = capturedAt
        }
    }

    /// Encode interleaved sRGB RGBA8 bytes into JPEG/HEIC data.
    ///
    /// - Parameters:
    ///   - exif: Optional EXIF metadata to embed. Nil omits the Exif dict entirely.
    static func encode(rgba8: [UInt8], width: Int, height: Int,
                       format: Format, quality: Double, exif: ExifMetadata? = nil) throws -> Data {
        // The CGContext below reads width*height*4 bytes from `rgba8`; a mismatch is an OOB read.
        guard width > 0, height > 0, rgba8.count == width * height * 4 else {
            throw ImageEncoderError.contextFailed
        }
        guard let cs = CGColorSpace(name: CGColorSpace.sRGB) else { throw ImageEncoderError.contextFailed }
        let bitmapInfo = CGImageAlphaInfo.noneSkipLast.rawValue | CGImageByteOrderInfo.order32Big.rawValue // RGBX, explicit byte order
        var bytes = rgba8
        guard let ctx = CGContext(data: &bytes, width: width, height: height,
                                  bitsPerComponent: 8, bytesPerRow: width * 4,
                                  space: cs, bitmapInfo: bitmapInfo),
              let cg = ctx.makeImage() else { throw ImageEncoderError.contextFailed }

        let out = NSMutableData()
        guard let dest = CGImageDestinationCreateWithData(
            out, format.utType.identifier as CFString, 1, nil)
        else { throw ImageEncoderError.destinationFailed }

        // Build the metadata properties dict.  The TIFF Software tag and ICC color profile
        // are written unconditionally; Exif dict only when exif != nil.
        //
        // JPEG TIFF dict note: ImageIO only returns `{TIFF}` from
        // CGImageSourceCopyPropertiesAtIndex for JPEG when the IFD0 contains the standard
        // resolution tags (XResolution/YResolution/ResolutionUnit).  Without those, a
        // Software-only IFD0 is silently swallowed on read-back.  We always emit resolution
        // so the TIFF dict is visible; Software round-trips correctly as a side-effect.
        var imageProps: [CFString: Any] = [
            kCGImageDestinationLossyCompressionQuality: quality,
            kCGImagePropertyTIFFDictionary: [
                "Software":       "Stack Stack Stack",
                "XResolution":    72,
                "YResolution":    72,
                "ResolutionUnit": 2,    // 2 = pixels/inch
            ] as NSDictionary,
            // Tag the output as sRGB so decoders interpret the pixel values correctly.
            // kCGImagePropertyProfileName sets the profile name in the image properties dict
            // (and in JFIF-style colorspace signaling); combined with kCGImagePropertyColorModel
            // it ensures the decoded CGImage's colorSpace.name returns "kCGColorSpaceSRGB"
            // (the Quartz constant for the named sRGB profile — same as real iPhone JPEG output).
            kCGImagePropertyColorModel: kCGImagePropertyColorModelRGB,
            kCGImagePropertyProfileName: "sRGB IEC61966-2.1",
        ]

        // Supply the raw ICC data bytes for formats that embed a full APP2 ICC block (e.g. HEIC).
        // For JPEG, ImageIO uses the profile name tag above; the raw ICC data is passed as a hint
        // but the "ICCProfile" key is not a documented constant.  Both paths produce a decoded
        // CGImage whose colorSpace.name is "kCGColorSpaceSRGB" — matching real iPhone JPEG output.
        // copyICCData() is available iOS 9+.
        if let iccData = cs.copyICCData() as Data? {
            imageProps["ICCProfile" as CFString] = iccData as CFData
        }

        if let exif {
            var exifDict: [CFString: Any] = [:]
            if let iso = exif.iso {
                exifDict[kCGImagePropertyExifISOSpeedRatings] = [Int(iso)]
            }
            if let shutter = exif.shutterSeconds {
                exifDict[kCGImagePropertyExifExposureTime] = shutter
            }
            if let date = exif.capturedAt {
                // EXIF DateTimeOriginal format is "yyyy:MM:dd HH:mm:ss" — uses a FIXED
                // en_US_POSIX locale so the colons in the date portion are never replaced by
                // locale-specific separators on non-Gregorian locales.  Current timezone is
                // correct here: the spec asks for local capture time, not UTC.
                let fmt = DateFormatter()
                fmt.locale = Locale(identifier: "en_US_POSIX")
                fmt.dateFormat = "yyyy:MM:dd HH:mm:ss"
                exifDict[kCGImagePropertyExifDateTimeOriginal] = fmt.string(from: date)
            }
            imageProps[kCGImagePropertyExifDictionary] = exifDict
        }

        CGImageDestinationAddImage(dest, cg, imageProps as CFDictionary)
        guard CGImageDestinationFinalize(dest) else { throw ImageEncoderError.finalizeFailed }
        return out as Data
    }
}

File: android/app/src/main/kotlin/com/jdpneto/stackstackstack/ImageEncoder.kt
package com.jdpneto.stackstackstack

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Color
import android.graphics.ColorSpace
import androidx.exifinterface.media.ExifInterface
import java.io.ByteArrayOutputStream
import java.util.Locale

/** Thrown when encoding fails. Mirrors [ImageEncoderError] in the iOS app. */
sealed class ImageEncoderError(msg: String) : Exception(msg) {
    object ContextFailed    : ImageEncoderError("Failed to create Bitmap context")
    object DestinationFailed: ImageEncoderError("Failed to create output stream")
    object FinalizeFailed   : ImageEncoderError("Failed to compress image")
}

/**
 * Encodes/decodes images to JPEG or HEIC (API 30+). Mirrors the iOS [ImageEncoder] exactly:
 * same format raw-values, same EXIF fields, same Software tag, same sRGB colour-space guarantee.
 */
object ImageEncoder {

    /**
     * `String`-backed: raw values are persisted (AppSettings + StackRecord.format) — renaming a
     * case silently breaks stored preferences and library records. Mirrors Swift's `rawValue`.
     */
    enum class Format(
        /** Stable persisted storage key — must NEVER change. Mirrors Swift's `rawValue`. */
        val rawValue: String
    ) {
        JPEG("jpeg"),
        HEIC("heic");

        /** The file extension library files use for this format. */
        val fileExtension: String
            get() = when (this) {
                JPEG -> "jpg"
                HEIC -> "heic"
            }

        /**
         * The [Bitmap.CompressFormat] corresponding to this encoder format.
         * HEIC is accessed via reflection because [Bitmap.CompressFormat.HEIC] is not present
         * in the platform SDK stub jar (it exists on-device from API 30+ but isn't exposed
         * through the compile-time stubs). On runtimes where HEIC is unavailable, compress()
         * will return false and [ImageEncoderError.FinalizeFailed] is thrown — the coordinator
         * falls back to JPEG (spec §3 honesty rule: HEIC is OEM-defined).
         */
        val bitmapCompressFormat: Bitmap.CompressFormat
            get() = when (this) {
                JPEG -> Bitmap.CompressFormat.JPEG
                HEIC -> heicCompressFormat ?: Bitmap.CompressFormat.JPEG  // fallback in Robolectric
            }

        companion object {
            /** Parse from a persisted raw value; null for unknown/corrupt. */
            fun fromRawValue(value: String?): Format? =
                entries.firstOrNull { it.rawValue == value }

            /**
             * [Bitmap.CompressFormat.HEIC] accessed via reflection (not in compile-time stubs).
             * Null when HEIC encoding is unavailable (e.g. Robolectric, older devices).
             */
            @Suppress("UNCHECKED_CAST")
            val heicCompressFormat: Bitmap.CompressFormat? by lazy {
                try {
                    val clazz = Bitmap.CompressFormat::class.java
                    clazz.getField("HEIC").get(null) as? Bitmap.CompressFormat
                } catch (e: Exception) { null }
            }
        }
    }

    /**
     * EXIF metadata written into the encoded result. All fields are optional; null = omit.
     * Mirrors [ExifMetadata] in the iOS app.
     */
    data class ExifMetadata(
        /** ISO speed (maps to [ExifInterface.TAG_ISO_SPEED_RATINGS]). */
        val iso: Double? = null,
        /** Exposure time in seconds (maps to [ExifInterface.TAG_EXPOSURE_TIME]). */
        val shutterSeconds: Double? = null,
        /** Capture timestamp (maps to [ExifInterface.TAG_DATETIME_ORIGINAL]).
         *  Stored as POSIX seconds (seconds since 1970-01-01 UTC). */
        val capturedAtPosix: Double? = null
    )

    /**
     * Encode interleaved sRGB RGBA8 bytes into JPEG/HEIC data.
     *
     * @param rgba8   Row-major RGBA bytes, length = width × height × 4.
     * @param width   Image width in pixels.
     * @param height  Image height in pixels.
     * @param format  Target format.
     * @param quality Compression quality 0.0–1.0 (maps to 0–100 for [Bitmap.compress]).
     * @param exif    Optional EXIF metadata to embed. Null omits the EXIF block.
     * @throws ImageEncoderError on any failure.
     */
    @Throws(ImageEncoderError::class)
    fun encode(
        rgba8: ByteArray,
        width: Int,
        height: Int,
        format: Format,
        quality: Double,
        exif: ExifMetadata? = null
    ): ByteArray {
        // Guard: buffer size must match dimensions exactly (OOB protection mirrors iOS).
        if (width <= 0 || height <= 0 || rgba8.size != width * height * 4) {
            throw ImageEncoderError.ContextFailed
        }

        // Build an ARGB_8888 Bitmap in sRGB from the RGBA bytes.
        // RGBA8 → ARGB_8888: Android's Bitmap stores ARGB in native byte order via Color.argb();
        // the engine's rgba8 has R at [0], G at [1], B at [2], A at [3] per pixel.
        val bitmap = try {
            Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888,
                /* hasAlpha */ false, ColorSpace.get(ColorSpace.Named.SRGB))
        } catch (e: Exception) {
            throw ImageEncoderError.ContextFailed
        }

        val pixels = IntArray(width * height)
        for (i in pixels.indices) {
            val base = i * 4
            val r = rgba8[base].toInt() and 0xFF
            val g = rgba8[base + 1].toInt() and 0xFF
            val b = rgba8[base + 2].toInt() and 0xFF
            // Alpha ignored (engine always writes fully-opaque RGBA).
            pixels[i] = Color.rgb(r, g, b)
        }
        bitmap.setPixels(pixels, 0, width, 0, 0, width, height)

        // Compress to the target format.
        val out = ByteArrayOutputStream()
        val q = (quality * 100).toInt().coerceIn(0, 100)
        val ok = bitmap.compress(format.bitmapCompressFormat, q, out)
        bitmap.recycle()
        if (!ok) throw ImageEncoderError.FinalizeFailed

        val compressed = out.toByteArray()

        // Embed EXIF + Software tag via ExifInterface (mutates the byte array in place).
        // Always write the Software tag; write the Exif block only when exif != null.
        return embedExif(compressed, format, exif)
    }

    /**
     * Embed EXIF metadata into an already-compressed JPEG/HEIC byte array via a temp file.
     * [ExifInterface.saveAttributes] requires file-backed mode to actually persist; stream-backed
     * mode is read-only. Returns the byte array with embedded metadata, or [data] unchanged if
     * EXIF embedding fails (EXIF failure must never lose the compressed image).
     */
    private fun embedExif(data: ByteArray, format: Format, exif: ExifMetadata?): ByteArray {
        return try {
            val tmp = java.io.File.createTempFile("sss_exif", ".${format.fileExtension}")
            try {
                tmp.writeBytes(data)
                val ei = ExifInterface(tmp.absolutePath)

                // Software tag — always written (mirrors iOS "Stack Stack Stack").
                ei.setAttribute(ExifInterface.TAG_SOFTWARE, "Stack Stack Stack")

                if (exif != null) {
                    exif.iso?.let { iso ->
                        ei.setAttribute(ExifInterface.TAG_ISO_SPEED_RATINGS, iso.toInt().toString())
                    }
                    exif.shutterSeconds?.let { shutter ->
                        // EXIF ExposureTime as a rational string "num/denom".
                        // For small shutter speeds, use a fraction; for >= 1s, use whole seconds.
                        ei.setAttribute(ExifInterface.TAG_EXPOSURE_TIME, doubleToExifRational(shutter))
                    }
                    exif.capturedAtPosix?.let { posix ->
                        // EXIF DateTimeOriginal format: "yyyy:MM:dd HH:mm:ss" in the local timezone.
                        // Matches the iOS FIXED en_US_POSIX locale formatting for the same field.
                        val sdf = java.text.SimpleDateFormat("yyyy:MM:dd HH:mm:ss", Locale.US)
                        sdf.timeZone = java.util.TimeZone.getDefault()
                        val date = java.util.Date((posix * 1000).toLong())
                        ei.setAttribute(ExifInterface.TAG_DATETIME_ORIGINAL, sdf.format(date))
                    }
                }

                ei.saveAttributes()
                tmp.readBytes()
            } finally {
                tmp.delete()
            }
        } catch (e: Exception) {
            // EXIF embedding failure must not lose the compressed image — return without EXIF.
            data
        }
    }

    /**
     * Convert a shutter-speed Double to an EXIF rational string "num/denom".
     * E.g. 0.02 → "1/50", 1.5 → "3/2".
     */
    private fun doubleToExifRational(value: Double): String {
        if (value <= 0) return "0/1"
        if (value >= 1.0) {
            val num = (value * 100).toLong()
            val den = 100L
            val g = gcd(num, den)
            return "${num / g}/${den / g}"
        }
        // Find a denominator with 1 as numerator for fast shutter speeds.
        val denom = (1.0 / value).toLong().coerceAtLeast(1)
        return "1/$denom"
    }

    private fun gcd(a: Long, b: Long): Long = if (b == 0L) a else gcd(b, a % b)
}
